src/main/java/io/confluent/connect/hdfs/TopicPartitionWriter.java

```diff
@@ -785,6 +801,40 @@ private void closeTempFile() {
     }
   }
 
+  private void fetchSchemaFromLatestFile() {
+    if (compatibility != StorageSchemaCompatibility.NONE && offset != -1) {
+      log.debug("Current schema is not set for partition {}, "
+          + "getting schema from file with latest offsets.", tp);
+      // check on WAL restored latest file to avoid recursive search
+      if (recovered && latestOffsetsFile != null) {
+        currentSchema = schemaFileReader.getSchema(
+            connectorConfig,
+            new Path(latestOffsetsFile)
+        );
+        log.trace("Retrieved schema from file with latest offset {}", latestOffsetsFile);
+      } else {
+        // if WAL restored file is N/A fallback on a recursive scan approach of filenames
+        log.debug("Could not find latest offsets file from WAL for schema initialization, "
+            + "searching for latest offsets file on filesystem.");
+        String topicDir = FileUtils.topicDirectory(url, topicsDir, tp.topic());
+        CommittedFileFilter filter = new TopicPartitionCommittedFileFilter(tp);
+        log.trace("Fetching file with max offset.");
+        FileStatus fileStatusWithMaxOffset = FileUtils.fileStatusWithMaxOffset(
+            storage,
+            new Path(topicDir),
+            filter
+        );
```

Comment:
we already do this in `readOffsets()`, can we further optimize and only do this once, maybe save the return value of `readOffsets()`?

Reply:
That is a good thought, I started attempting that when I realized the issue is the following:

Both `readoffsets()` and `fetchSchemaFromLatestFile()` are under the `write()` method. 

The `write()` method only calls `readOffsets()` in some cases, before the state is `WRITE_STARTED`. 
However, `write()` may call `fetchSchemaFromLatestFile()` on an execution when `readoffsets()` wasn't called. 

This means that the stored value for the max file status that we get from `readoffsets()` may be from an earlier (outdated) call to `write()`. In that case the max file needs to be recomputed here. 

This also made me realize it holds for the case of the readOffsetsFromWAL() as well. We can't rely on that saved file, so we'll need to extract from the WAL again here. I will update the PR.

Reply:
I updated the PR with the revision for this issue.

Reply:
On a second thought, the schema recovery logic should be followed up in a separate PR.